torchbenchmark/models/soft_actor_critic/__init__.py

```diff
@@ -0,0 +1,243 @@
+    action_batch = action_batch.to(device)
+    reward_batch = reward_batch.to(device)
+    done_batch = done_batch.to(device)
+
+    agent.train()
+    ###################
+    ## CRITIC UPDATE ##
+    ###################
+    alpha = torch.exp(log_alpha)
+    with torch.no_grad():
+        action_dist_s1 = agent.actor(next_state_batch)
+        action_s1 = action_dist_s1.rsample()
+        logp_a1 = action_dist_s1.log_prob(action_s1).sum(-1, keepdim=True)
+        target_action_value_s1 = torch.min(
+            target_agent.critic1(next_state_batch, action_s1),
+            target_agent.critic2(next_state_batch, action_s1),
+        )
+        td_target = reward_batch + gamma * (1.0 - done_batch) * (
+            target_action_value_s1 - (alpha * logp_a1)
+        )
+
+    # update critics
+    agent_critic1_pred = agent.critic1(state_batch, action_batch)
+    agent_critic2_pred = agent.critic2(state_batch, action_batch)
+    td_error1 = td_target - agent_critic1_pred
+    td_error2 = td_target - agent_critic2_pred
+    critic_loss = 0.5 * (td_error1 ** 2 + td_error2 ** 2)
+    if per:
+        critic_loss *= imp_weights
+    critic_loss = critic_loss.mean()
+    critic_optimizer.zero_grad()
+    critic_loss.backward()
+    if critic_clip:
+        torch.nn.utils.clip_grad_norm_(
+            chain(agent.critic1.parameters(), agent.critic2.parameters()), critic_clip
+        )
+    critic_optimizer.step()
+
+    if update_policy:
+        ##################
+        ## ACTOR UPDATE ##
+        ##################
+        dist = agent.actor(state_batch)
+        agent_actions = dist.rsample()
+        logp_a = dist.log_prob(agent_actions).sum(-1, keepdim=True)
+        actor_loss = -(
+            torch.min(
+                agent.critic1(state_batch, agent_actions),
+                agent.critic2(state_batch, agent_actions),
+            )
+            - (alpha.detach() * logp_a)
+        ).mean()
+        actor_optimizer.zero_grad()
+        actor_loss.backward()
+        if actor_clip:
+            torch.nn.utils.clip_grad_norm_(agent.actor.parameters(), actor_clip)
+        actor_optimizer.step()
+
+        ##################
+        ## ALPHA UPDATE ##
+        ##################
+        alpha_loss = (-alpha * (logp_a + target_entropy).detach()).mean()
+        log_alpha_optimizer.zero_grad()
+        alpha_loss.backward()
+        log_alpha_optimizer.step()
+
+    if per:
+        new_priorities = (abs(td_error1) + 1e-5).cpu().detach().squeeze(1).numpy()
+        buffer.update_priorities(priority_idxs, new_priorities)
+
+class Model(BenchmarkModel):
+    task = REINFORCEMENT_LEARNING.OTHER_RL
+    def __init__(self, device=None, jit=False):
+        super(Model, self).__init__()
+        self.device = device
+        self.jit = jit
+        self.args = SACConfig()
+        # Construct agent
+        current_dir = os.path.dirname(os.path.abspath(__file__))
+        self.train_env = load_gym(self.args.env_id, self.args.seed)
```

Comment:
wait, are you actually using the gym for this? i thought you were going to mock the env

Reply:
I cut the "simulation boundary" at mujoco, not gym. This model uses gym but doesn't use mujoco, so I don't mock the env here. Similarly in https://github.com/pytorch/benchmark/pull/387, I mock the mujoco environment but still use gym.spaces. Please let me know whether this is reasonable, or you suggest we mock both mujoco and gym, and get rid of the gym dependency.

Reply:
I think if you can make a clear argument for why you set that boundary, so that others can follow the precedent in the future, I am ok with it.